spec/core/ics-026-routing-module/UPGRADES.md

```diff
@@ -0,0 +1,141 @@
+  counterpartyChannelIdentifier: Identifier,
+  version: string) => (version: string, err: Error) {
+    // defined by the module
+}
+```
+
+#### **OnChanUpgradeTry**
+
+`onChanUpgradeTry` will verify the upgrade-chosen parameters and perform custom `TRY` logic. 
+If the upgrade-chosen parameters are invalid, the callback must return an error to abort the handshake. 
+If the counterparty-chosen version is not compatible with this modules
+supported versions, the callback must return an error to abort the handshake. 
+If the versions are compatible, the try callback must select the final version
+string and return it to core IBC.
+`onChanUpgradeTry` may also perform custom initialization logic.
+
+`onChanUpgradeTry` is also responsible for making sure that the application is recoverable to its pre-upgrade state. The application may either store any new metadata in separate paths, or store the previous metadata under a different path so it can be restored.
+
+```typescript
+function onChanUpgradeTry(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  counterpartyVersion: string) => (version: string, err: Error) {
+    // defined by the module
+}
+```
+
+#### **OnChanUpgradeAck**
+
+`onChanUpgradeAck` will error if the counterparty selected version string
+is invalid to abort the handshake. It may also perform custom ACK logic.
+
+After `onChanUpgradeAck` returns, the application upgrade is complete on this end so any 
+auxilliary data stored for the purposes of recovery is no longer needed and may be deleted.
+
+The application MUST have its state fully migrated to start processing packet data according to the new application parameters by the time the callback returns.
+
+```typescript
+function onChanUpgradeAck(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier, 
+  counterpartyVersion: string) {
+    // defined by the module
+}
+```
+
+#### **OnChanUpgradeConfirm**
+
+`onChanUpgradeConfirm` will perform custom CONFIRM logic. It MUST NOT error since the counterparty has already approved the handshake, and transitioned to using the new upgrade parameters.
+
+After `onChanUpgradeConfirm` returns, the application upgrade is complete so any 
+auxilliary data stored for the purposes of recovery is no longer needed and may be deleted.
+
+The application MUST have its state fully migrated to start processing packet data according to the new application parameters by the time the callback returns.
+
+```typescript
+function onChanUpgradeConfirm(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier) {
+    // defined by the module
+}
+```
+
+#### **OnChanUpgradeRestore**
+
+`onChanUpgradeRestore` will be called on `cancelChannelUpgrade` and `timeoutChannelUpgrade` to restore the application to its pre-upgrade state.
```

Comment:
`onChanUpgradeRestore` is also called by `restoreChannel()`, which may be called from `chanUpgradeTry` and `chanUpgradeAck`